tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.cc

```diff
@@ -0,0 +1,160 @@
+      return false;
+    }
+    auto select_list = dynamic_cast<const zetasql::ASTSelectList*>(node);
+
+    // Count how many columns have the input name
+    int count = 0;
+
+    for (auto column : select_list->columns()) {
+      if (column_name == get_column_name(column)) {
+        count++;
+      }
+    }
+
+    // count > 1 means duplicate columns.
+    if (count > 1) {
+      return true;
+    }
+
+  };
+
+  auto candidate = find_node(&node, predicator);
+  return dynamic_cast<const zetasql::ASTSelectList*>(candidate);
+
+}
+
+std::string get_column_name(const zetasql::ASTSelectColumn* column_node) {
+  if (column_node == nullptr) {
+    return "";
+  }
+
+  // If alias is set, return the alias
+  auto alias = column_node->alias();
+  if (alias != nullptr) {
+    return alias->identifier()->GetAsString();
+  }
+
+  for (int i = 0; i < column_node->num_children(); i++) {
+    auto child = column_node->child(i);
+    if (child->node_kind() == zetasql::ASTNodeKind::AST_ALIAS) {
+      continue;
+    }
+
+    auto path_expression = dynamic_cast<const zetasql::ASTPathExpression*>(child);
+
+    // If the child is of ASTSelectColumn is not ASTPathExpression, then the name should be empty
+    if (path_expression == nullptr) {
+      return "";
+    }
+
+    // Return the last name of ASTPathExpression as the column name.
+    return path_expression->last_name()->GetAsString();
+  }
+}
+
+void replace_duplicate_columns_of_select_list(
+    const zetasql::ASTSelectList& select_list,
+    absl::string_view column_name,
+    zetasql_base::UnsafeArena* arena,
+    zetasql::IdStringPool* id_string_pool
+) {
+  std::vector<zetasql::ASTSelectColumn*> duplicate_columns;
+
+  // Find all the duplicate columns
+  for (auto column : select_list.columns()) {
+    if (column_name == get_column_name(column)) {
+      duplicate_columns.push_back(const_cast<zetasql::ASTSelectColumn*>(column));
+    }
+  }
+
+  for (int i = 0; i < duplicate_columns.size(); i++) {
+    auto new_alias = absl::StrCat(column_name, "_", std::to_string(i + 1));
+    // rename the column with 1-based index
+    update_alias(duplicate_columns[i], new_alias, arena, id_string_pool);
+  }
+}
+
+void update_alias(
+    zetasql::ASTSelectColumn* column_node,
+    absl::string_view new_alias,
+    zetasql_base::UnsafeArena* arena,
+    zetasql::IdStringPool* id_string_pool
+) {
+  if (column_node->alias() == nullptr) {
+    auto alias_node = create_alias_node(new_alias, arena, id_string_pool);
+    column_node->AddChild(alias_node);
+    ((zetasql::ASTNode*) column_node)->InitFields();
+    return;
+  }
+
+  auto identifier = const_cast<zetasql::ASTIdentifier*>(column_node->alias()->identifier());
+  identifier->SetIdentifier(id_string_pool->Make(new_alias));
+  ((zetasql::ASTNode*) identifier)->InitFields();
+}
+
+zetasql::ASTAlias* create_alias_node(
+    absl::string_view new_alias,
+    zetasql_base::UnsafeArena* arena,
+    zetasql::IdStringPool* id_string_pool
```

Comment:
are we expecting to change arena and id_string_pool in the function body? if not mark const and pass by reference

Reply:
I believe we do. We use arena to allocate space and String pool to create a new string.